### PR TITLE
Uratpoly

### PR DESCRIPTION
**Pull request by srajangarg**
SymEngine rational polys added
Yet to add piranha and flint rat polys, and all tests

Also some functions like `is_zero`, `is_mul` etc. have been removed temporarily. I will add them for all classes later on.

### CODE REVIEW

File: symengine/expand.cpp [hunk 1]
```diff
@@ -297,6 +298,26 @@ class ExpandVisitor : public BaseVisitor<ExpandVisitor>
             _coef_dict_add_term(multiply, r);
             return;
         }
+        if (is_a<Integer>(*self.get_exp()) && is_a<URatPoly>(*_base)) {
```
Comment (isuruf) [inline symengine/expand.cpp]: Can you remove this? Polynomials are not still integrated into `add`, `mul` and `pow`, so let's deal with later. (We shouldn't have added the lines above as well.)

File: symengine/mp_class.h [hunk 2]
```diff
@@ -60,6 +60,18 @@ inline integer_class operator"" _z(const char *str)
...
+{
+    return rational_class(integer_class(str));
+}
+}
+
+inline void mp_tdiv_qr(rational_class &q, rational_class &r,
+                       const rational_class &a, const rational_class &b)
+{
+    q = a / b;
+    r = 0;
```
Comment (isuruf) [inline symengine/mp_class.h]: What's the use of this function? Why do you need `r`?

File: symengine/mp_class.h [hunk 3]
```diff
@@ -129,6 +141,16 @@ inline void mp_pow_ui(integer_class &res, const integer_class &i,
...
 }
 
+inline void mp_pow_ui(rational_class &res, const rational_class &i,
+                      unsigned long n)
+{
+    integer_class tmp;
+    mpz_pow_ui(tmp.get_mpz_t(), i.get_num().get_mpz_t(), n);
+    mpq_set_num(res.get_mpq_t(), tmp.get_mpz_t());
+    mpz_pow_ui(tmp.get_mpz_t(), i.get_den().get_mpz_t(), n);
+    mpq_set_den(res.get_mpq_t(), tmp.get_mpz_t());
```
Comment (isuruf) [inline symengine/mp_class.h]: `tmp` is not needed here, right?

File: symengine/mp_class.h [hunk 4]
```diff
@@ -276,8 +298,8 @@ inline auto get_mpz_t(const piranha::integer &i) -> decltype(i.get_mpz_view())
     return i.get_mpz_view();
 }
 
-inline void mp_pow_ui(piranha::integer &res, const piranha::integer &i,
-                      unsigned long n)
+template <typename T>
+inline void mp_pow_ui(T &res, const T &i, unsigned long n)
```
Comment (isuruf) [inline symengine/mp_class.h]: Can you repeat this function without using templates so that only `piranha::integer` and `piranha::rational` can be used?

File: symengine/polys/basic_conversions.h [hunk 5]
```diff
@@ -120,7 +120,7 @@ class BasicToUPolyBase : public BaseVisitor<V>
     {
         integer_class i = x.i;
         if (i != 0)
-            dict = P::container_from_dict(gen, {{0, i}});
+            dict = P::cont_from_dict(gen, {{0, i}});
```
Comment (isuruf) [inline symengine/polys/basic_conversions.h]: Can you remove this change?

File: symengine/polys/uexprpoly.h [hunk 6]
```diff
@@ -7,7 +7,8 @@
 
 #include <symengine/expression.h>
 #include <symengine/monomials.h>
-#include <symengine/polys/upolybase.h>
+// #include <symengine/polys/upolybase.h>
```
Comment (isuruf) [inline symengine/polys/uexprpoly.h]: Why this change?

File: symengine/polys/uexprpoly.h [hunk 7]
```diff
@@ -186,8 +188,8 @@ class UExprPoly : public UPolyBase<UExprDict, UExprPoly>
                                           map_int_Expr &&d);
     static RCP<const UExprPoly> from_vec(const RCP<const Basic> &var,
                                          const std::vector<Expression> &v);
-    static UExprDict container_from_dict(const RCP<const Basic> &var,
-                                         map_int_Expr &&d);
+    static UExprDict cont_from_dict(const RCP<const Basic> &var,
```
Comment (isuruf) [inline symengine/polys/uexprpoly.h]: Remove this change

File: symengine/expand.cpp (hunk 1 above)
Comment (srajangarg) [inline symengine/expand.cpp]: So, remove the `expand` test cases?

File: symengine/mp_class.h (hunk 2 above)
Comment (srajangarg) [inline symengine/mp_class.h]: required in `divides_upoly`. Was already being used by the integer variant by calling this and checking remainder.

File: symengine/expand.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/expand.cpp]: Yes

File: symengine/polys/uexprpoly.h (hunk 6 above)
Comment (srajangarg) [inline symengine/polys/uexprpoly.h]: `UExprPoly` now uses the implementations provided by `usymenginepoly.h`

File: symengine/printer.h [hunk 8]
```diff
@@ -33,8 +33,12 @@ class Precedence : public BaseVisitor<Precedence>
     template <typename Poly>
     void bvisit_upoly(const Poly &x)
     {
-        if (x.get_dict().size() == 1) {
-            auto it = x.get_dict().begin();
+        auto begin = x.begin();
+        auto it = x.begin();
+        auto end = x.end();
+
+        if (end == ++begin) {
```
Comment (isuruf) [inline symengine/printer.h]: Why not have `x.size() == 1` ?

File: symengine/printer.h [hunk 9]
```diff
@@ -61,11 +65,27 @@ class Precedence : public BaseVisitor<Precedence>
         bvisit_upoly(x);
     }
 
+    void bvisit(const URatPoly &x)
```
Comment (isuruf) [inline symengine/printer.h]: Instead of repeating this everywhere, you can do,
```
template <Container, Poly>
bvisit(const UPolyBase<Container, Poly> &x)
{
    bvisit_upoly(static_cast<const Poly &>(x));
}
```

File: symengine/printer.cpp [hunk 10]
```diff
@@ -414,30 +416,34 @@ void uintpoly_print(const T &x, std::ostringstream &s)
...
         s << "0";
+    return s.str();
 }
 
 void StrPrinter::bvisit(const UIntPoly &x)
 {
-    std::ostringstream s;
-    uintpoly_print(x, s);
-    str_ = s.str();
+    str_ = upoly_print<UIntPoly>(x);
```
Comment (isuruf) [inline symengine/printer.cpp]: Same here as in `bvisit_upoly`

File: symengine/printer.h (hunk 8 above)
Comment (srajangarg) [inline symengine/printer.h]: The "polynomial size" is defined as `0` if it's a zero poly otherwise `degree+1`  (as used by flint)
This was the only way I can think of to make it work for all the upolys.

File: symengine/printer.cpp [hunk 11]
```diff
@@ -362,15 +363,16 @@ void StrPrinter::bvisit(const GaloisField &x)
 
 // UIntPoly printing, tests taken from SymPy and printing ensures
```
Comment (isuruf) [inline symengine/printer.cpp]: Change this to say `Integer` or `Rational` coefficient polynomial printing

File: symengine/printer.cpp [hunk 12]
```diff
@@ -362,15 +363,16 @@ void StrPrinter::bvisit(const GaloisField &x)
...
+std::string upoly_print(const P &x)
 {
+    std::ostringstream s;
     // bool variable needed to take care of cases like -5, -x, -3*x etc.
     bool first = true;
     // we iterate over the map in reverse order so that highest degree gets
     // printed first
     for (auto it = x.obegin(); it != x.oend(); ++it) {
-        integer_class m = to_integer_class(it->second);
+        typename P::coef_type m = it->second;
```
Comment (isuruf) [inline symengine/printer.cpp]: `auto` can be used here

File: symengine/printer.cpp (hunk 10 above)
Comment (srajangarg) [inline symengine/printer.cpp]: thanks

File: symengine/printer.h (hunk 8 above)
Comment (isuruf) [inline symengine/printer.h]: Okay, then don't use `size` at all.
```
if (x.end() == ++x.begin()) {   // If there is only one term
....
} else if (x.end() == x.begin() { // If equal to 0
...
}
```

File: symengine/mp_class.h (hunk 2 above)
Comment (isuruf) [inline symengine/mp_class.h]: You don't have to use this kind of hacks. It's better to write two functions for `divides_upoly`

File: symengine/printer.cpp (hunk 10 above)
Comment (srajangarg) [inline symengine/printer.cpp]: I will actually have to shift the `upoly_print` code to the `.h` file? Should I go ahead?

File: symengine/polys/usymenginepoly.h [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    static RCP<const Poly> from_vec(const RCP<const Basic> &var,
+                                    const std::vector<Cf> &v)
+    {
+        return make_rcp<const Poly>(var, Container::from_vec(v));
+    }
+
+    static Container cont_from_dict(const RCP<const Basic> &var,
+                                    std::map<Key, Cf> &&d)
```
Comment (isuruf) [inline symengine/polys/usymenginepoly.h]: `std::map<Key, Cf>` should just be `Container` right?

File: symengine/polys/usymenginepoly.h [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+    std::vector<Cf> multieval(const std::vector<Cf> &v) const
+    {
+        // this is not the optimal algorithm
+        std::vector<Cf> res(v.size());
+        for (Key i = 0; i < v.size(); ++i)
+            res[i] = eval(v[i]);
+        return res;
+    }
+
+    inline const std::map<Key, Cf> &get_dict() const
```
Comment (isuruf) [inline symengine/polys/usymenginepoly.h]: Same here

File: symengine/polys/usymenginepoly.h [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+        return this->poly_.dict_;
+    }
+
+    inline Cf get_coeff(Key x) const
+    {
+        return this->poly_.get_coeff(x);
+    }
+
+    typedef typename std::map<Key, Cf>::const_iterator iterator;
+    typedef typename std::map<Key, Cf>::const_reverse_iterator r_iterator;
```
Comment (isuruf) [inline symengine/polys/usymenginepoly.h]: Same here

File: symengine/polys/usymenginepoly.h [hunk 16]
```diff
@@ -0,0 +1,163 @@
+#ifndef SYMENGINE_USYMENGINEPOLY_H
+#define SYMENGINE_USYMENGINEPOLY_H
+
+#include <symengine/polys/upolybase.h>
+
+namespace SymEngine
+{
+
+template <typename Container, template <typename X, typename Y> class BaseType,
+          typename Poly, typename Key>
```
Comment (isuruf) [inline symengine/polys/usymenginepoly.h]: You can deduce `Key` using `Container` right?

File: symengine/polys/usymenginepoly.h (hunk 13 above)
Comment (srajangarg) [inline symengine/polys/usymenginepoly.h]: No, `Container` is for eg. `UIntDict` and not the underlying map

File: symengine/polys/usymenginepoly.h (hunk 16 above)
Comment (srajangarg) [inline symengine/polys/usymenginepoly.h]: Agreed

**Comment by srajangarg** [comment]
@isuruf ping

**Comment by isuruf** [comment]
+1
